### PR TITLE
Direct PyTorch to OpenVINO IR conversion API

### PR DESCRIPTION
**Pull request by dkurt**
Proposal for direct PyTorch to OpenVINO IR conversion procedure (ONNX free).

Key aspects:
* PyTorch native serialization formats (`torch.save`) can save only the weights (state_dict) or topology definition without connections and dependency with origin modules. So command line way is not applicable

Example usage (AlexNet from `torchvision`):
```python
import numpy as np
import torch
import torch.nn as nn
from torch.autograd import Variable
import torchvision.models as models

np.random.seed(324)
torch.manual_seed(32)

# Create model
model = models.alexnet(pretrained=True)

# Run model to get reference output
inp = Variable(torch.randn([1, 3, 227, 227]))
model.eval()
ref = model(inp)

# Convert to OpenVINO IR
import mo_pytorch
mo_pytorch.convert(model, input_shape=[1, 3, 227, 227], model_name='alexnet')

# Run model with OpenVINO and compare outputs
from openvino.inference_engine import IECore

ie = IECore()
net = ie.read_network('alexnet.xml', 'alexnet.bin')
exec_net = ie.load_network(net, 'CPU')
out = exec_net.infer({'input': inp.detach().numpy()})
out = next(iter(out.values()))
print('diff:', np.max(np.abs(out - ref.detach().numpy())))  # 1.3113022e-06
```

supported models:
* `torchvision.models.alexnet`
* `torchvision.models.resnet18`
* `torchvision.models.segmentation.deeplabv3_resnet50`
* [Detectron2 RetinaNet](https://github.com/facebookresearch/detectron2/blob/master/configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml) (https://github.com/dkurt/openvino/pull/2)

### CODE REVIEW

File: model-optimizer/extensions/ops/MatMul.py [hunk 1]
```diff
@@ -162,10 +162,6 @@ def infer(node: Node):
         Raises on any shape inconsistency
         """
         name = node.soft_get('name', str(node.id))
-        connected_in_ports = {idx: port for idx, port in node.in_ports().items() if not port.disconnected()}
```
Comment (jane-intel) [inline model-optimizer/extensions/ops/MatMul.py:165]: Please do not remove this lines. They are valid. If PyTorch has MatMul that has different semantic -- we should normalize it separately on the front phase

File: model-optimizer/extensions/load/pytorch/loader.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        graph.graph['layout'] = 'NCHW'
+
+        update_extractors_with_extensions(pytorch_op_extractors)
+
+        # Create a dummy input
+        inp = Variable(torch.randn(list(argv.placeholder_shapes)))
+
+        model = argv.input_model
+
+        def myhook(self, inputs, output):
```
Comment (jane-intel) [inline model-optimizer/extensions/load/pytorch/loader.py:50]: nice naming)

File: model-optimizer/extensions/front/pytorch/parameter_ext.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    enabled = True
+
+    @classmethod
+    def extract(cls, node):
+        # t_type = node.pb.type.tensor_type
+        # attrs = {
+        #     'shape': np.array([d.dim_value for d in t_type.shape.dim], dtype=np.int64),
+        #     'data_type': TENSOR_TYPE_TO_NP_TYPE[t_type.elem_type]
+        # }
+        Parameter.update_node_stat(node, {})
```
Comment (jane-intel) [inline model-optimizer/extensions/front/pytorch/parameter_ext.py:31]: we should set shape and dtype for Parameter -- it is essential for the op

File: model-optimizer/extensions/front/pytorch/linear.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    enabled = True
+
+    def replace_op(self, graph: Graph, node: Node):
+        shape = Const(graph, {'value': [0, -1]}).create_node()
+        flatten = Reshape(graph, dict(name=node.in_node(0).name + '/flatten')).create_node([node.in_node(0), shape])
+
+        inputs = [flatten, node.in_node(1)]
+        if len(node.in_nodes()) > 2:
+            inputs.append(node.in_node(2))  # bias
+        matmul = MatMul(graph, dict(name=node.name, transpose_b=True)).create_node(inputs)
```
Comment (jane-intel) [inline model-optimizer/extensions/front/pytorch/linear.py:36]: MatMul doesn't perform bias addition, please insert Add node after MatMul and everything will be fine)

File: model-optimizer/extensions/front/pytorch/pooling_ext.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        return cls.enabled
+
+
+class AdaptiveAvgPool2dFrontExtractor(FrontExtractorOp):
+    op = 'AdaptiveAvgPool2d'
+    enabled = True
+
+    @classmethod
+    def extract(cls, node):
+        Identity.update_node_stat(node)
```
Comment (jane-intel) [inline model-optimizer/extensions/front/pytorch/pooling_ext.py:67]: :)

**Review by jane-intel — COMMENTED** [review]
To trigger CI cycle, MO unit-tests should pass, for that please update automation/BOM.txt file

**Comment by slyalin** [comment]
@dkurt, can you comment on advantages that this approach gives in comparison to the ONNX exporting?

File: model-optimizer/extensions/load/pytorch/loader.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            global layer_id
+            layer_type = self.__class__.__name__
+
+            # Create a unique name
+            name = '{}_{}'.format(layer_type, layer_id)
```
Comment (jane-intel) [inline model-optimizer/extensions/load/pytorch/loader.py:55]: we have `unique_id` method for that in `Graph` object

File: model-optimizer/extensions/load/pytorch/loader.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            layer_id += 1
+
+            graph.add_node(name, kind='op', op=layer_type, name=name, shape=list(output.shape), module=self)
```
Comment (jane-intel) [inline model-optimizer/extensions/load/pytorch/loader.py:58]: shape setting for intermediate tensors is not a good practice. we infer them during `partial_infer` stage

File: model-optimizer/extensions/load/pytorch/loader.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    def load(self, graph: Graph):
+        argv = graph.graph['cmd_params']
+        graph.graph['fw'] = 'pytorch'
+        graph.graph['layout'] = 'NCHW'
+
+        update_extractors_with_extensions(pytorch_op_extractors)
+
+        # Create a dummy input
+        inp = Variable(torch.randn(list(argv.placeholder_shapes)))
```
Comment (jane-intel) [inline model-optimizer/extensions/load/pytorch/loader.py:46]: Not scale-able solution.

After the loading we extract original Placeholder shapes.
In case user defines new shapes, InputCut transformation performs the shape renewal.

But here we need input shapes for the Loader. So we force user to always define them during MO call.

Multi input topologies are not covered.

File: model-optimizer/extensions/ops/MatMul.py (hunk 1 above)
Comment (dkurt) [reply model-optimizer/extensions/ops/MatMul.py:165]: It's about `DecomposeBias` logic. I need to create a fully connected layer with bias.

File: model-optimizer/extensions/load/pytorch/loader.py (hunk 8 above)
Comment (dkurt) [reply model-optimizer/extensions/load/pytorch/loader.py:46]: PyTorch loader requires to specify input shape (as in demo) or input tensor (need to think what's better) because we perform forward pass inside.

```python
mo_pytorch.convert(model, input_shape=[1, 3, 227, 227])
```

> But here we need input shapes for the Loader. So we force user to always define them during MO call.

`input_shape` in this case is MO flag

File: model-optimizer/extensions/front/pytorch/pooling_ext.py (hunk 5 above)
Comment (dkurt) [reply model-optimizer/extensions/front/pytorch/pooling_ext.py:67]: Yeah, sorry about that :) Works fine for AlexNet but will definitely replace.

File: model-optimizer/extensions/ops/MatMul.py (hunk 1 above)
Comment (dkurt) [reply model-optimizer/extensions/ops/MatMul.py:165]: Reverted, thanks!

File: model-optimizer/extensions/load/pytorch/loader.py (hunk 7 above)
Comment (dkurt) [reply model-optimizer/extensions/load/pytorch/loader.py]: removed

**Comment by rblaczkowski** [comment]
Please merge master to your branch to fetch the newest OpenVino ONNX CI state.

**Comment by dkurt** [comment]
see https://github.com/openvinotoolkit/openvino_contrib/tree/master/modules/mo_pytorch
